### PR TITLE
Blog post: Run Minitest on Github Actions with parallel jobs using build matrix

### PR DESCRIPTION
**Pull request by ArturT**

### CODE REVIEW

File: _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+author: "Artur Trzop"
+categories: continuous_integration
+og_image: "/images/blog/posts/run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix/github-minitest.jpeg"
+---
+
+How to run Ruby on Rails tests in Minitest on Github Actions? What to do if tests are slow? How to manage complex workflows? You can use Github Actions build matrices to divide Minitest files between jobs and run the test suite much faster.
+
+<img src="/images/blog/posts/run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix/github-minitest.jpeg" style="width:300px;margin-left: 15px;float:right;" alt="Minitest, Ruby, Github, Github Actions" />
+
+If your Minitest tests are taking dozen minutes and you would like to save some time for your Ruby engineering team then you could use tests parallelization on your CI server.
```
Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:14]: "dozen minutes" means literally "12 minutes".
You could say "dozens of minutes" if you mean to say multiples of 12-minute periods (similarly to how in Polish we would say "dziesiątki" or "kilkadziesiąt")

File: _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+To run tests as fast as possible you need to split them into equal buckets (into parallel jobs). But how to do it? Some of the test files can be super fast to execute, other Minitest files can take minutes if they run system tests (E2E tests).
+
+There is also an aspect of preparing the test environment for each parallel job. By preparing I mean you need to clone a repository, install ruby gems or load them from a cache, maybe you need to load some docker container, etc. This can take various times on each parallel job. Random network errors happen like network delay to [load cached gems](/2021/how-to-load-ruby-gems-from-cache-on-github-actions), or maybe Github Actions from time to time will start one of your jobs late comparing to others. It's an inevitable issue in the network environment and can cause your tests to run a different amount of time on each parallel job. This is visible on the graph below and it causes the CI build to be slower.
```
Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:18]: `various times` means something else. maybe "various amounts of time" would fit here?

Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:18]: `start one of your jobs late comparing to others`
-> `compared to others` or `relative to others`

Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:18]: `to run a different amount of time` -> `to run for a different amount of time`

File: _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+<img src="/images/blog/posts/how-to-speed-up-ruby-and-javascript-tests-with-ci-parallelisation/not-optimal-tests-split.png" style="width:100%;" alt="not optimal tests split on CI server, CI parallelism" />
+
+In a perfect scenario you would like to cover all those problems and no matter what still be able to split Minitest work in parallel jobs in a way that ensures the tests on each parallel job completes at a similar time. This guarantees no bottlenecks. The perfect tests split is on the below graph.
```
Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:22]: `all those problems` -> `all these problems` (I _think_)

File: _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+## Split tests in a dynamic way with Queue Mode
+
+You can use [Knapsack Pro](https://knapsackpro.com/?utm_source=docs_knapsackpro&utm_medium=blog_post&utm_campaign=run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix) Queue Mode to split tests in a dynamic way between parallel jobs. This way each job consumes tests from a queue until the queue is empty. Simply speaking this allows you to utilize your CI server resources efficiently and run tests in optimal time.
+
+I described [how Queue Mode splits Ruby and JavaScript tests in parallel with a dynamic test suite split](/2020/how-to-speed-up-ruby-and-javascript-tests-with-ci-parallelisation). You can learn from that article about it.
+
+## Github Actions build matrix to run parallel tests
+
+Github Actions has [build matrix feature](https://docs.github.com/en/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix) that allows running many jobs at the same time. You can use it to run your Minitest tests between parallel jobs.
```
Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:34]: has a build matrix feature

File: _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          KNAPSACK_PRO_FIXED_QUEUE_SPLIT: true
+          KNAPSACK_PRO_LOG_LEVEL: info
+        run: |
+          bundle exec rake knapsack_pro:queue:minitest
+{% endraw %}
+{% endhighlight %}
+
+## Summary
+
+As you can see the slow Minitest test suite is no longer an issue for you. QA, Testers, or Automation Engineers can benefit from improving the CI build speed and allowing their software developers team to deliver products faster. You can learn more at [Knapsack Pro](https://knapsackpro.com/?utm_source=docs_knapsackpro&utm_medium=blog_post&utm_campaign=run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix).
```
Comment (shadre) [inline _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md:121]: unsure about the "is no longer an issue for you" part.

Maybe it's better to say something like "doesn't need be an issue for you". Then in the next sentence we could change "can benefit" to "could benefit".

**Review by shadre — APPROVED** [review]
nicely done :+1: 

left some ideas

File: _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md (hunk 5 above)
Comment (shadre) [reply _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md]: @ArturT sorry I made a mistake in the suggestion. This should read "doesn't need _to_ be an issue for you" instead.

Comment (ArturT) [reply _posts/2021-04-06-run-minitest-on-github-actions-with-parallel-jobs-using-build-matrix.md]: Ok. Fixed in https://github.com/KnapsackPro/docs.knapsackpro.com/commit/36dc0e6928e1d8e3887a14269d333d2a137ea609

**Pull request merged**